README.md

```diff
@@ -1,45 +1,263 @@
+    unchanged_keys = batch1.keys - keys_whose_score_changed_in(batch1,batch2)
+    redis.zrem(queue_key,*members_of(unchanged_keys))   # critical section end
+
+Gerald changed it so a second snapshot of the cold end of the queue is
+taken after process_batch_slowly().  Only documents whose timestamps
+did not change between the two snapshots are removed from the queue.
+
+Notice how the critical section no longer includes
+process_batch_slowly().  Instead it only spans two Redis ops and some
+local set arithmetic which.
+
+The critical section and the Forgotten Dirtiness Problem which it
+causes is still there, but is much smaller.  In practice we have
+process_batch_slowly() taking minutes, but even in extreme situations
+this critical section never took more than 3 seconds.
+
+## Proposal: The Ick Pattern
+
+In October 2015, while reviewing the Forgotten Dirtiness problem, we
+identified the Hot Data Starvation Problem.  We developed Ick and
+switched to this almost familiar pattern:
+
+    # in any process: whenever a document is dirtied
+    Ick.ickadd(redis,queue_key,Time.now.to_f,document_id)
+
+    # in the indexer process:
+    batch = Ick.ickreserve(redis,queue_key,batch_size)
+    process_batch_slowly(batch)
+    # burn down the queue only if the batch succeeded
+    Ick.ickcommit(redis,queue_key,*members_of(batch))   # critical section gone
+
+
+Ick solves for failover via a two phase commit protocol between
+**ickreserve** and **ickcommit**.  If there is a failure during
+process_batch_slowly(batch), the next time time we call **ickreserve**
+we will just get the same batch - it will have resided unchanged in
+the consumer set until we get happy and call **ickcommit**.
+
+Ick solves the Forgotten Dirtiness Problem by virtue of
+**ickreserve**’s implicit atomicity and the fact that **ickcommit** is
+only ever called from the indexer and producers do not mutate the
+consumer set.
+
+Ick solves the Hot Data Starvation Problem by a subtle change in
+ICKADD.  Unlike ZADD, which overwrites the old score when a message is
+re-added, or ZADD NX which always preserves the old score, ICKADD
+always takes the _min_ of the old and new scores.  Thus, Ick tracks
+the first-known ditry time for a message even when there is time skew
+in the producers.  The longer entries stay in the consumer set, the
+more they implicitly percolate toward the cold end regardless of how
+many updates they receive.  Ditto in the consumer set.  Provided that
+all producers make a best effort to use only current or future
+timestamps when they call ICKADD, the ICKRESERVE batch will always
+include the oldest entries and there will be no starvation.
+
+Apology: I know that [Two-Phase
+Commit](https://en.wikipedia.org/wiki/Two-phase_commit_protocol) has a
+different technical meaning than what Ick does.  Unfortunately I can't
+find a better name for this very common failsafe queue pattern.  I
+suppose we could the Redis sorted set as the coordinator and the
```

Comment:
> I suppose we could the Redis sorted set

Missing a verb there, I think.

Reply:
Thanks, added it "I suppose we could consider the Redis ..."